### PR TITLE
[jit] reenable rand_like fusion when there is no broadcast

### PR DESCRIPTION
**Pull request by ngimel**
Reenables rand_like fusion if no tensor is broadcasted in the fusion group. This is a sufficient but not necessary condition for fused rand_like to produce correct results, and it has an unpleasant side effect of falling back to non-fused path if rand_like was optimistically included in the fusion group, but there is a broadcast in the fusion group not necessarily related to rand_like. E.g. before this PR, if the network had (biasAdd -> relu -> dropout), fuser could fuse biasAdd and relu, now it will try fusing the whole thing (if dropout is expressed via rand_like) and fall back every time.

### CODE REVIEW

File: torch/csrc/jit/fuser/executor.cpp
Comment (zou3519) [inline torch/csrc/jit/fuser/executor.cpp]: The naming for this is confusing, it took me a few minutes to figure out what was going on. Could you rename this to `shouldExpandArgs` (and also keep the `expandArgs` function?)

File: torch/csrc/jit/fuser/executor.cpp
Comment (zou3519) [inline torch/csrc/jit/fuser/executor.cpp]: nit: maybe `dry_run` is a better name for `should_expand`, it's a little weird hearing "should_expand" for a function "expandArgs" that implies it always expands its arguments.

**Review by zou3519 — APPROVED** [review]
lgtm, but I had some minor comments about naming

**Review by zou3519 — APPROVED** [review]
thanks!

**Comment by apaszke** [comment]
Well, as you've noted in many cases it will actually not result in good fusions, so why are we adding this back like this? Maybe it's better to have a more robust fallback path?

**Comment by ngimel** [comment]
I agree that in some cases the lack of flexible fallback will result in regression, because optimistic fusion failed. As a side note, similar thing would happen now with at::cat that would be optimistically fused, and then, if cat inputs are of the different sizes, which is a pretty common case, the whole fusion will fail and fallback to nonfused implementation. 
Still, I think it would be good to start enabling random fusions as in this PR, and gradually work on improving fallback. One way of minimizing regressions would be to copy input arguments as is done in #14957 https://github.com/pytorch/pytorch/pull/14957/files#diff-9f801fd31b2860a98a07b168f805e621R303. Then we can fail during codegen and still cleanly run fallback. Without copying inputs, expandArgs modifies inputs in place, and once expandArgs is run no failure is allowed. I was concerned about copying inputs, but may be the overhead of this is not too bad, and it's required anyway for #14957 . If we are allowed to fail during codegen, then we can fail only on cases that really require random broadcast, and only the fusions preceding or following random broadcast would be regressed, which is a corner case. biasAdd -> relu ->dropout would be handled correctly.  
If this is preferred, I can wait until #14957 is landed and take another stab after that.

File: torch/csrc/jit/fuser/executor.cpp
Comment (apaszke) [inline torch/csrc/jit/fuser/executor.cpp]: nit: this uses an indentation of 3 spaces which is pretty weird

**Review by apaszke — APPROVED** [review]

**Pull request closed without merging**
